beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs

```diff
@@ -0,0 +1,1769 @@
+            critical_lock.store_keys.insert(root);
+        }
+
+        drop(critical_lock);
+        drop(maintenance_lock);
+        Ok(())
+    }
+
+    /// maintain the cache
+    pub fn do_maintenance(&self, cutoff_epoch: Epoch) -> Result<(), AvailabilityCheckError> {
+        // ensure memory usage is below threshold
+        let threshold = self.capacity * 3 / 4;
+        self.maintain_threshold(threshold, cutoff_epoch)?;
+        // clean up any keys on the disk that shouldn't be there
+        self.prune_disk(cutoff_epoch)?;
+        Ok(())
+    }
+
+    /// Enforce that the size of the cache is below a given threshold by
+    /// moving the least recently used items to disk.
+    fn maintain_threshold(
+        &self,
+        threshold: usize,
+        cutoff_epoch: Epoch,
+    ) -> Result<(), AvailabilityCheckError> {
+        // ensure only one thread at a time can be deleting things from the disk or
+        // moving things between memory and storage
+        let maintenance_lock = self.maintenance_lock.lock();
+
+        let mut stored = self.critical.read().in_memory.len();
+        while stored > threshold {
+            let read_lock = self.critical.upgradable_read();
+            let lru_entry = read_lock
+                .in_memory
+                .peek_lru()
+                .map(|(key, value)| (*key, value.clone()));
+
+            let (lru_root, lru_pending_components) = match lru_entry {
+                Some((r, p)) => (r, p),
+                None => break,
+            };
+
+            if lru_pending_components
+                .epoch()
+                .map(|epoch| epoch < cutoff_epoch)
+                .unwrap_or(true)
+            {
+                // this data is no longer needed -> delete it
+                let mut write_lock = RwLockUpgradableReadGuard::upgrade(read_lock);
+                write_lock.in_memory.pop_entry(&lru_root);
+                stored = write_lock.in_memory.len();
+                continue;
+            } else {
+                drop(read_lock);
+            }
+
+            // write the lru entry to disk (we aren't holding any critical locks while we do this)
+            self.overflow_store
+                .persist_pending_components(lru_root, lru_pending_components)?;
+            // now that we've written to disk, grab the critical write lock
+            let mut write_lock = self.critical.write();
+            if let Some((new_lru_root_ref, _)) = write_lock.in_memory.peek_lru() {
+                // need to ensure the entry we just wrote to disk wasn't updated
+                // while we were writing and is still the LRU entry
+                if *new_lru_root_ref == lru_root {
+                    // it is still LRU entry -> delete it from memory & record that it's on disk
+                    write_lock.in_memory.pop_entry(&lru_root);
+                    write_lock.store_keys.insert(lru_root);
+                    stored = write_lock.in_memory.len();
```

Comment:
Nice catch! Yes this should be moved to right above `drop(write_lock)`. As currently written there's a rare case which would cause this to over-prune:

1) critical.read().in_memory.len() = threshold + 1
2) While we are writing LRU entry to disk, the entry is updated
3) Update makes entry complete, removing it from cache so it's imported into forkchoice
4) critical.read().in_memory.len() = threshold -> loop should terminate but it continues because `stored` isn't updated

As far as I can tell this would be ultimately harmless but it should be corrected

Reply:
Updated in 655f82e50f9d68aef8e2da2e6ffc464c089c61f3